algebra-core/src/biginteger/macros.rs

```diff
@@ -198,6 +198,48 @@ macro_rules! bigint_impl {
 
                 res
             }
+
+            #[inline]
+            fn mul_no_reduce(this: &[u64], other: &[u64]) -> Self {
+                debug_assert!(this.len() <= $num_limbs / 2);
+                debug_assert!(this.len() == other.len());
+
+                let mut r = [0u64; $num_limbs];
+                for i in 0..$num_limbs / 2 {
+                    let mut carry = 0u64;
+                    for j in 0..$num_limbs / 2 {
+                        r[j + i] =
+                            arithmetic::mac_with_carry(r[j + i], this[i], other[j], &mut carry);
+                    }
+                    r[$num_limbs / 2 + i] = carry;
+                }
+                Self::new(r)
+            }
+
+            #[inline]
+            fn mul_no_reduce_lo(this: &[u64], other: &[u64]) -> Self {
+                debug_assert!(this.len() == $num_limbs);
+                debug_assert!(this.len() == other.len());
+
+                let mut r = [0u64; $num_limbs];
+                for i in 0..$num_limbs {
+                    let mut carry = 0u64;
+                    for j in 0..($num_limbs - i) {
+                        r[j + i] =
+                            arithmetic::mac_with_carry(r[j + i], this[i], other[j], &mut carry);
+                    }
+                }
+                Self::new(r)
+            }
+
+            #[inline]
+            fn from_slice(slice: &[u64]) -> Self {
```

Comment:
Do we want to return an error when the slice is too large?

Reply:
Might actually help the compiler by not doing bounds check for every slice access

Reply:
Oh I was referring to mul_no_reduce for bounds checking. For this one, I guess it makes no difference. It ought to be clear how many limbs a biginteger type has so it's easy to diagnose when it's suitable or not.

Reply:
I've discovered that when one can return an `Option`, it's probably the right thing to do (eg: with `from_repr`)